server/server.go

```diff
@@ -151,15 +182,23 @@ func NewServer(cfg *config.Config, driver IDriver) (*Server, error) {
 	}
 
 	var err error
-	if cfg.Socket != "" {
-		if s.listener, err = net.Listen("unix", cfg.Socket); err == nil {
-			log.Infof("Server is running MySQL Protocol through Socket [%s]", cfg.Socket)
-		}
-	} else {
+
+	if s.cfg.Host != "" && s.cfg.Port != 0 {
 		addr := fmt.Sprintf("%s:%d", s.cfg.Host, s.cfg.Port)
 		if s.listener, err = net.Listen("tcp", addr); err == nil {
 			log.Infof("Server is running MySQL Protocol at [%s]", addr)
 		}
+		if cfg.Socket != "" {
+			go s.forwardUnixSocketToTCP(cfg.Socket, addr) // listen on both
+		}
+	} else if cfg.Socket != "" {
+		if s.listener, err = net.Listen("unix", cfg.Socket); err == nil {
+			log.Infof("Server is running MySQL Protocol through Socket [%s]", cfg.Socket)
+		} else {
+			log.Fatalf("err: %s", err) // in use?
```

Comment:
Yes, all errors must be handled!

Instead of logging fatally, this function should probably return the error and log at the error level as needed.

Reply:
I've refactored this now, so the err is caught further down.  Outcome is the same.